### PR TITLE
[Constraint solver] Small steps toward unifying solver entry points

### PR DESCRIPTION
**Pull request by DougGregor**
Start hollowing out the code path in the constraint solver that does all of the setup/application for the top-level expression differently from everything else.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test compiler performance

File: lib/Sema/CSSolver.cpp [hunk 1]
```diff
@@ -1115,80 +1115,93 @@ bool ConstraintSystem::solve(Expr *&expr,
...
-      .highlight(expr->getSourceRange());
-    return true;
-  }
+      if (Options.contains(
+            ConstraintSystemFlags::AllowUnresolvedTypeVariables)) {
+        auto ambiguousSolutions = std::move(solution).takeAmbiguousSolutions();
+        solutions.assign(std::make_move_iterator(ambiguousSolutions.begin()),
+                         std::make_move_iterator(ambiguousSolutions.end()));
+        dumpSolutions();
+        solution.markAsDiagnosed();
```
Comment (CodaFi) [inline lib/Sema/CSSolver.cpp]: This is a use-after-move.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/Sema/CSSolver.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/CSSolver.cpp]: Good catch, thank you!

**Pull request merged**